### DISCUSSION TITLE
Maturin imports not resolved

### DISCUSSION THREAD

**Discussion by Haadi-Khan** [microsoft/pyright#8145] [General]
**Describe the bug**
Pyright fails to import packages from Maturin/PyO3 (Rust bindings for python). I verified that this was a fault of pyright by trying 3 different editors which all rely on the pyright language server (Emacs, Neovim, and Zed). The import works fine using Pylance and the standard python ishell. Other imports within the venv behave normally (i.e. external packages and internal python files).

**Code or Screenshots**
Pyright (Emacs)
<img width="875" alt="image" src="https://github.com/microsoft/pyright/assets/44217975/51d5bf1e-527c-4ac9-b5d3-5684ee8d448f">
>  Pyright [reportAttributeAccessIssue]: "Anyon" is unknown import symbol

Pylance (VSCode)
<img width="705" alt="image" src="https://github.com/microsoft/pyright/assets/44217975/f7b8e826-4b33-4289-9e31-a4d1c4196d80">

**VS Code extension or command-line**
pyright 1.1.366
Run via pyright-langserver --stdio

**Additional**
I think this bug is fixable by adding the appropriate file path to a pyproject.toml/pyrightconfig.json, but this should be a default behavior of pyright given that pylance "just works".

Thanks in advance

**Comment by erictraut** [comment]
This is unlikely to be a bug in pyright, but I can help you try to figure out a solution. I'll need additional information. First, is this the [maturin 1.6.0](https://pypi.org/project/maturin/) library in pypi (i.e. `pip install maturin`)? Second, can you provide a short code snippet (in text form) that demonstrates the problem? You have some screen shots in your report above that appear to be importing from a package named `anyon_braiding_simulator`, but I don't know what this has to do with the `maturin` package. When I install `maturin`, I don't see any other package installed by the name of `anyon_braiding_simulator`, so I'm not sure where that comes from.

**Comment by Haadi-Khan** [comment]
Thanks for the quick reply

Yes, this is the maturin from pypi.

My project is titled `anyon_braiding_simulator`, so when I run `maturin develop` to build the bindings, it creates a python module titled `anyon_braiding_simulator` which contains said bindings.

for a code snippet:
```python
import pytest
from anyon_braiding_simulator import Anyon, IsingTopoCharge


def test_anyon():
    anyon = Anyon('thisisatest', IsingTopoCharge.Psi, (1, 1))
    assert anyon.name == 'thisisatest'
    assert anyon.charge == IsingTopoCharge.Psi
    assert anyon.position == (1, 1)
```

This is a simple test file I have. The code runs fine, but autocomplete is broken for Anyon/IsingTopoCharge, and the same error is listed as my initial post. My project structure is identical to the [mixed layout](https://www.maturin.rs/project_layout#mixed-rustpython-project) if that helps with configuring the `pyproject.toml` includes

**Comment by erictraut** [comment]
OK, so these are generated bindings. I obviously don't have access to your `anyon_braiding_simulator` module. Is your project checked in to a public repo? 

Is the `anyon_braiding_simulator` module local to your project or installed in your python environment (i.e. in `site_packages`? Is it a ".py" or ".pyi" (stub) file?

**Comment by Haadi-Khan** [comment]
Here's the [project](https://github.com/Cornell-QCA/anyon-braiding-simulator/tree/fusion)

The module is installed within my venv's site packages. It has an `__init__.py` file which calls `anyon_braiding_simulator.cpython-312-darwin.so` (located in the same dir)

**Comment by erictraut** [comment]
Can you provide step-by-step instructions for reproducing your problem? Assume I know nothing about the `maturin` library or your project (because I don't :) ). Are you installing your `anyon_braiding_simulator` using an editable install?

**Comment by Haadi-Khan** [comment]
Sorry, I should've given more info about using `maturin`

I'll be more or less paraphrasing [this](https://www.maturin.rs/tutorial), so feel free to read it if my explanation is insufficient.

After installing `maturin` w/ pip, running `maturin new` and writing the rust code, you build and install the bindings for python by running `maturin develop`. Immediately after running this command, you're able to access the bindings in a python shell/editor by doing `import <project_name>`. In my case it's `anyon_braiding_simulator`. The bindings are installed in the venv's `site-packages/<project_name>` directory.

The main project layout is structured to have `src/` for rust, and `python/<project_name>` and `python/tests` for python. I'm calling the bindings generated from `maturin develop` from both python directories via `from anyon_braiding_simulator import ...`. This is valid python, and the python interpreter has no issues with it (nor does pylance).

It may be important to note that in the `target/wheels` directory, there's a `.whl` file to install `anyon_braiding_simulator` with `pip`, but it's unnecessary for the interpreter to work.

However, pyright can't seem to find the import paths for the packages. Notably, pyright gets confused by the name `anyon_braiding_simulator` because it's also the name of a directory within `python/`, and pyright's suggested files are members of that directory (but the python interpreter behaves as I expect, and not as pyright does, i.e. it calls the rust bindings).

Hopefully this helps for figuring out a solution. I'm rather new to `maturin`, so I'm sorry if my explanation is lacking.

**Comment by erictraut** [comment]
I'm looking for a specific set of command-line instructions, something like:
1. Clone the [project](https://github.com/Cornell-QCA/anyon-braiding-simulator/tree/fusion)
2. Create a virtual environment using Python 3.12 and activate it
3. Run `pip install maturin`
4. etc.

I'm going to convert this from a bug report to a discussion topic since this is unlikely to be a bug in pyright and likely to be a configuration issue or a problem with the generated bindings.

**Comment by Haadi-Khan** [comment]
First 3 instructions are the same. Then:
4. `maturin develop`
5. open any of the test files in any pyright-enabled editor (let's say `vim python/tests/test_state.py`)
6. Observe the error `"Anyon" is an unknown import symbol` (and likewise for all other imported classes

I am certain this is not an issue with the code because:
1. The code compiles and functions as expected
2. The import behaves normally using the Pylance language server in VSCode

I assume this can be fixed with some inclusion for the pyright import path in `pyproject.toml`

**Comment by bruno-robert** [comment]
I think this is because Matuin and pyo3 don't generate `pyi` files. You can write your own and add it to the Maturin project directory, see [adding-python-type-information](https://www.maturin.rs/project_layout#adding-python-type-information).

Once the `pyi` file(s) is(are) in place, pyright works correctly.

There is an [open issue](https://github.com/PyO3/pyo3/issues/2454) tracking adding automatic python interface file generation to pyo3.
